Jna error code and plugin usage

Piankero opened the pull request:
## Basics

Check relevant points but **please do not remove entries**.
Do not describe the purpose of this PR in the PR description but:

- [x] Short descriptions should be in the release notes (added as entry in
      `doc/news/_preparation_next_release.md` which contains `_(my name)_`)
      **Please always add something to the the release notes.**
- [ ] Longer descriptions should be in documentation or in design decisions.
- [x] Describe details of how you changed the code in commit messages
      (first line should have `module: short statement` syntax)
- [ ] References to issues, e.g. `close #X`, should be in the commit messages.

## Checklist

Check relevant points but **please do not remove entries**.
For docu fixes, spell checking, and similar none of these points below
need to be checked.

- [x] I added unit tests
- [x] I ran all tests locally and everything went fine
- [x] affected documentation is fixed
- [ ] I added code comments, logging, and assertions (see [Coding Guidelines](https://master.libelektra.org/doc/CODING.md))
- [ ] meta data is updated (e.g. README.md of plugins and [METADATA.ini](https://master.libelektra.org/doc/METADATA.ini))

## Review

Reviewers will usually check the following:

- [x] Code is conforming to [our Coding Guidelines](https://master.libelektra.org/doc/CODING.md)
- [ ] APIs are conforming to [our Design Guidelines](https://master.libelektra.org/doc/DESIGN.md)
- [x] Code is consistent to [our Design Decisions](https://master.libelektra.org/doc/decisions)
- [ ] Documentation is concise and good to read
- [x] Examples are well chosen and understandable

## Additional Notes

I completely overhauled the jna binding and also removed the strange code examples for a plugin since it did nothing useful than logging stuff to the console without any interaction to elektra. The new Plugin class is actually usable.

markus2330 commented:
Thank you for the PR! I'll look at it after the release. Please make it pass the build server.

markus2330 commented on src/bindings/jna/libelektra4j/src/main/java/org/libelektra/plugin/PropertiesStorage.java:
```diff
@@ -1,74 +0,0 @@
-package org.libelektra.plugin;
```
Why is this plugin removed here? The PR is vey difficult to read due to reformatting of all files. Please describe some formatting for Java in doc/CODING.md (can be separate to Groovy, or also reformat scripts/jenkins/Jenkinsfile)

Piankero replied:
I did not reformat anything. I deleted a lot of files which basically had no senseful meaning like the Echo.java or the PropertiesStorage.java because:

1. I want to use the native elektra plugin from is implemented by the new Plugin class and used by the ErrorCodeTest (I use the `error` plugin there)
2. This would be in a conflicting name with the current `Plugin` interface which more or less just gave 4 methods to implement and basically had no real connection to KDB at all
3. The "implemented Plugins" in Java did not interact with KDB at all except for reading something from the parent key. Isn't it the meaning of elektra to load all into a keyset and handle the operations there rather than doing it with the native java code that directly manipulates the files? The specification for example could never be used with a java plugin. 

Due to all of these reasons I decided to remove them because they might even mislead new developers in implementing plugins in java

markus2330 replied:
> I did not reformat anything.

Please look at the diff in GitHub. E.g. src/bindings/jna/libelektra4j/src/main/java/org/libelektra/Plugin.java had tabs before and in this PR it now uses spaces.

> The "implemented Plugins" in Java did not interact with KDB at all except for reading something from the parent key. Isn't it the meaning of elektra to load all into a keyset and handle the operations there rather than doing it with the native java code that directly manipulates the files? The specification for example could never be used with a java plugin.

These plugins are actual Elektra plugins which can be mounted (and used as storage plugin). Please read src/plugins/jni/README.md to understand the plugin interface.

I am very glad if you improve the plugin interface (even beyond the introduction of the new error concept) but:

- there needs to be documentation on how to use it
- you need to update existing documentation (this PR does not touch documentation at all)
- there need to be examples
- there need to be test cases

Piankero replied:
> Please look at the diff in GitHub. E.g. src/bindings/jna/libelektra4j/src/main/java/org/libelektra/Plugin.java had tabs before and in this PR it now uses spaces

On my GitHub I do not see a difference between tabs and spaces, do you have a special plugin or setting which shows this?

Apart from that: It is the default setting in intellij is to use spaces and it sometimes even suggests to use them instead of tabs if it sees it is used in a file. It is in general much more common to use spaces in java. I can revert that back to tabs if you like but imho it does not make a lot of difference at all

> These plugins are actual Elektra plugins which can be mounted (and used as storage plugin). Please read src/plugins/jni/README.md to understand the plugin interface.

I readded them and renamed the plugin to nativeplugin which was used by myself to for the test case.

markus2330 replied:
You see tabs/spaces by trying to mark them.

It is commons sense that mixing tabs and spaces is the worst of all possibilities. So yes, please revert to tabs. We can convert everything to spaces in a later PR to have more Java-style. In any case, it is good if this is documented in doc/CODING.md. Even better would be to automatically reformat with a tool (and on the build server).

Piankero replied:
> You see tabs/spaces by trying to mark them.

Honestly I have seen problem with line endings but never with tabs/spaces. Nontheless I think there is a good plugin in case you are using firefox or chrome for that https://github.com/glebm/render-whitespace-on-github. I used it now to quickly see tabs/spaces.

I fixed the tab/spaces stuff. Please review the rest of the PR :)

Piankero commented:
@markus2330 I have absolutely no idea why on that one build it does not find the error plugin and returns null. I will exclude let the test pass in case the plugin returns null

markus2330 commented on doc/news/_preparation_next_release.md:
```diff
@@ -99,6 +99,9 @@ you up to date with the multi-language support provided by Elektra.
 ### Java
 
 - Upgraded maven dependencies for java binding _(Michael Zronek)_
+- Completely overhauled the java binding to be able to use Elektra plugins directly. For an example see the [test case](src/bindings/jna/libelektra4j/src/test/java/org/libelektra/ErrorCodeTest.java). _(Michael Zronek)_
```
```suggestion
- Completely overhauled the Java binding to be able to use Elektra plugins directly. For an example see the [test case](src/bindings/jna/libelektra4j/src/test/java/org/libelektra/ErrorCodeTest.java). _(Michael Zronek)_
```

markus2330 commented on doc/news/_preparation_next_release.md:
```diff
@@ -99,6 +99,9 @@ you up to date with the multi-language support provided by Elektra.
 
 - Upgraded maven dependencies for java binding _(Michael Zronek)_
+- Completely overhauled the java binding to be able to use Elektra plugins directly. For an example see the [test case](src/bindings/jna/libelektra4j/src/test/java/org/libelektra/ErrorCodeTest.java). _(Michael Zronek)_
+- The java binding now supports the [error codes](doc/decisions/error_codes.md) in a native way. All exceptions contain the necessary information. For
```
```suggestion
- The Java binding now supports the [error codes](doc/decisions/error_codes.md) in a native way. All exceptions contain the necessary information. For
```

markus2330 commented on doc/news/_preparation_next_release.md:
```diff
@@ -99,6 +99,9 @@ you up to date with the multi-language support provided by Elektra.
 - Upgraded maven dependencies for java binding _(Michael Zronek)_
+- Completely overhauled the java binding to be able to use Elektra plugins directly. For an example see the [test case](src/bindings/jna/libelektra4j/src/test/java/org/libelektra/ErrorCodeTest.java). _(Michael Zronek)_
+- The java binding now supports the [error codes](doc/decisions/error_codes.md) in a native way. All exceptions contain the necessary information. For
+  an example please take a look at the [test case](src/bindings/jna/libelektra4j/src/test/java/org/libelektra/ErrorCodeTest.java). _(Michael Zronek)_
```
Twice the same example?

Piankero replied:
Yes, one shows how to use the `NativePlugin` and one shows how Exceptions are used. They are both  used in the test

markus2330 replied:
It would be good to have more than one example. And examples should not try to to much. I still do not understand what the differences in the new plugin system is.

If it is better, we should migrate the old plugins (which are only small examples anyway) to the new system.

Piankero replied:
I added another example in the README.md
I migrated the old plugins to the new Interface with added methods. Also NativeElektraPlugins implements it and works.
I removed Template&ExceptionTest as it is not really valuable

markus2330 commented on src/bindings/jna/libelektra4j/src/main/java/org/libelektra/NativeElektraPlugin.java:
```diff
@@ -0,0 +1,118 @@
+import java.util.List;
+
+
+public class NativeElektraPlugin {
```
No documentation for this class and its methods

markus2330 commented on src/bindings/jna/libelektra4j/src/main/java/org/libelektra/exception/ConflictingStateException.java:
```diff
@@ -0,0 +1,16 @@
+package org.libelektra.exception;
```
These files should be generated from the specification.

Piankero replied:
I fear that autogeneration is not possible.
The main reason is the class hierarchy which maps to the error code.

I have also mapped invisible classes (eg. InstallationException) which do not appear at all in the specification file. These classes are neccessary though so developers can catch the more generic exception which was one of the main intentions behind the error code design.

Furthermore the autogeneration would need to know which error code is below which for correct inheritence. So there would be error prone recursive parsing of the individual digits.
Additionally we would completely deny ourself special implementation in subclasses, e.g., the ConflictingState/ OutOfMemory Exception which have more of a special meaning. All classes would be equally with autogeneration.

markus2330 replied:
> I fear that autogeneration is not possible.

Then the specification is not like it should be. We want the specification useful for several languages.

> All classes would be equally with autogeneration.

It is always possible to specialize, like I also suggested in  #2697

Piankero replied:
The specification is useful for all languages. It is just that the specification *file* does not have all errors.
Autogeneration of C code to an exact equivalent optimized java code is very error prone. Especially when it comes to inheritance. There are actually [companies](https://www.mtsystems.com/) who specialize in this matter and still are not able to do it perfectly. Even from java to kotlin or vice versa it does not work well and they both rely on the jvm.

Implementing your desired autogeneration would be worth at least another bachelor thesis.

markus2330 replied:
> The specification is useful for all languages.

This is to be shown by you.

> It is just that the specification file does not have all errors.

Then how can it be useful?

> Especially when it comes to inheritance.

You maybe need to add the relevant information to the specification.

> Autogeneration [...]

You are talking something very offtopic here and use wrong terms. Nobody requested to write a transpiler.

The code (the exceptions) should be generated from the (one and only) specification src/error/specification

> Implementing your desired autogeneration would be worth at least another bachelor thesis.

If this would be the case then something somewhere went terribly wrong. In one bachelor thesis two complete bindings were written. The new specification should make it easier, not harder.

We also need to write some generator for Go and Rust.

Piankero replied:
> This is to be shown by you.

It is shown here that it works for both java and C very well which can be seen in this PR. It is just that autogeneration is very complicated and error prone.

> Then how can it be useful?

Because developers are only permitted to use those error codes which are relevant and not those which are too generic. On could also write down additional codes with a metatag that says it is not allowed to use but only for "documentation" but this is really just unnecessarily bloating of the specification file.

> The code (the exceptions) should be generated from the (one and only) specification src/error/specification

Lets talk about this in another elektra meeting as it will simply be too time intensive writing here all the time.

> If this would be the case then something somewhere went terribly wrong. In one bachelor thesis two complete bindings were written. The new specification should make it easier, not harder.
We also need to write some generator for Go and Rust.

Did these bindings include autogeneration and exception mapping for a language which has the same sort of inheritance as java? Go and Rust have a different form of inheritence which is easier to model

Piankero replied:
Lets talk about this in an elektra meeting as this would otherwise be more time intensive than needed

markus2330 replied:
Just to clarify what I am talking about: The specification in src/error/specification should be complete enough, that the relation (or subtyping, inheritance, ..) between the error codes can easily be known within a tool that generates exception classes (or needs to know if one error is subtype of another). If you find the encoding as error code letters too difficult yourself, why not encode it in an easier form?

> Did these bindings include autogeneration and exception mapping for a language which has the same sort of inheritance as java?

Yes, it used code generation far beyond of what we are talking here. Generating these few exception classes is trivial compared to that. Feel free to have a look at src/bindings/swig/python/kdb.i

> Lets talk about this in an elektra meeting as this would otherwise be more time intensive than needed

We can also talk about it but we also should document how we came to the conclusion.

And it is an "urgent" matter because:

1. you want to finish
2. the decision is needed for Go and Rust bindings which are finished soon, too

I started  #2871.

Piankero replied:
Since this is discussed in #2871 it can stay now as it is

markus2330 commented on src/bindings/jna/libelektra4j/src/main/java/org/libelektra/exception/KDBException.java:
```diff
@@ -0,0 +1,74 @@
+	}
+
+	@Override
+	public String getMessage() {
```
Shouldn't that print the whole message?

Piankero replied:
This is usually just 1-2 sentences and not a big bunch of additional information. It is a bit weird to try to marry C kind of error with java exceptions. I personally would only want to get the Reason message and try to avoid the other verbose and eventually unneccessary output. I can though print everything if you want.

markus2330 replied:
It is not about what I want but it is about consistency of error messages. Why should a Java-GUI print different error messages than the Qt-GUI?

Piankero replied:
The QT-GUI also just prints the Reason message and on expansion prints everything. I do not see why they are different

markus2330 replied:
I expected that "getReason" will return the reason and "getMessage" will return "everything".

Piankero replied:
getReason is elektra specific while getMessage is java specific. I simply wanted to have both working. But I will change it "everything"

markus2330 replied:
All these discussions would not be needed if you had some documentation about what these methods are supposed to do. It is completely unclear what "I simply wanted to have both working" means. What is a working method? One that does not throw exceptions?

Piankero replied:
If someone comes from C (or more specific Elektra) and uses the java binding he would expect "getReason" to be the method with the most "useful" String in it.
If however someone comes from Java he would expect the "getMessage()" method to be the most relevant one.
I wanted to supply both so either of them is satisfied. But you wish for the whole message in the reason part which is also perfectly fine

markus2330 replied:
I expected the getMessage() to be the most relevant one. So I wondered why it did not contain the whole error message.

> I wanted to supply both so either of them is satisfied.

Of course it makes sense to have getters for the individual parts but it also makes sense to have a getter for the standardized message format as specified by you in doc/decisions/error_message_format.md

Piankero replied:
You can see an example message now in the README.md

markus2330 replied:
Thank you!

But the methods are still not documented.

markus2330 commented on src/bindings/jna/libelektra4j/src/main/java/org/libelektra/exception/mapper/ExceptionMapperService.java:
```diff
@@ -0,0 +1,49 @@
+
+		String errorCode = k.getMeta("error/number").getString();
+
+		if (errorCode.equals(ResourceException.errorCode())) {
```
Also this here should be generated.

Piankero replied:
See above

markus2330 commented on src/bindings/jna/libelektra4j/src/main/java/org/libelektra/exception/mapper/ExceptionMapperService.java:
```diff
@@ -0,0 +1,49 @@
+		if (k.getMeta("error/number").isNull() && k.getMeta("warnings/#00/number").isNull()) {
+			Key temporaryError = Key.create("user/temporary/errorkey");
+			temporaryError.setMeta("error/number", InternalException.errorCode());
+			temporaryError.setMeta("error/reason", "Sorry, Elektra returned an error/warning but no " +
```
Undocumented hack to rewrite the reason? Why not append the keys when returning the message?

Piankero replied:
> Undocumented hack to rewrite the reason?

It is not to "rewrite" the reason but to actually give one because Elektra did not provide one. E.g., you make a call to Elektra, it returns -1 but the errorkey is empty. This should never happen in reality but I provided a fallback so the user would not receive a NullPointer/ Empty Message

Piankero replied:
If you wish I can add another constructor which expects (String errorCode, String reason) as paramters

markus2330 replied:
I do not think that another constructor solves the problem.

> you make a call to Elektra, it returns -1 but the errorkey is empty.

Then this is the wrong place to fix this. This code must be within the `kdb*` functions.

Piankero replied:
This is not "fix" per se but a safeguard. There may always be changes in the kdb functions which may cause no error to be returned. The java binding would at least catch this error and not return a null pointer. I already had this behavior when trying to load a plugin which was not existent on the build server.

Piankero replied:
> I do not think that another constructor solves the problem.

The additional constructor would be just some better usability like a constructor with a predefined size I have done in the KeySet class

markus2330 replied:
https://de.wikipedia.org/wiki/Design_by_contract

Clients should be able to rely on that they get an error message if they get `-1`. If this is not the case, we have a bug, please report it.

Piankero replied:
I know that it is design by contract. If the contract is violated though it is still good working code. I will simply write "please report a bug to libelektra..." instead.

markus2330 replied:
> I know that it is design by contract. If the contract is violated though it is still good working code.

Then please also describe your interface according design-by-contract. I did not see that you try to avoid to return null pointers.

> I will simply write "please report a bug to libelektra..." instead.

Yes, better.

Piankero replied:
> Yes, better.

Actually I see that the current error message is already in this form (did not see it in the marked PR line immediately) `Sorry, Elektra returned an error/warning but no information along it, please report this incident at https://issues.libelektra.org/`

markus2330 commented:
Thank you! Looks much better now :sparkle: The class design is much nicer.

Without docu, however, it is really hard to understand how a developer can use your classes.

markus2330 commented on src/bindings/jna/libelektra4j/src/main/java/org/libelektra/Plugin.java:
```diff
@@ -14,7 +22,7 @@
 	 * @return the plugin's return value for open
 	 */
-	int open(KeySet conf, Key errorKey);
+	int kdbOpen(KeySet conf, Key errorKey);
```
Why do you rename here? The prefix seems unnecessary (both for Plugin and PluginLoader).

It would be great if you can add a test where a Java plugin is loaded and tested.

Piankero replied:
The names are the equivalents to those in the C written implementation. I wanted it to be consistent but if you wish I can refactor it.

Piankero replied:
> It would be great if you can add a test where a Java plugin is loaded and tested.

I do not have written those plugins and also do not have the time to test them unfortunately.

markus2330 replied:
> The names are the equivalents to those in the C written implementation. I wanted it to be consistent but if you wish I can refactor it.

It is astonishing that such important decisions are not mentioned at all. **Please** document such things, it would have saved a lot of time for both of us.

src/libs/tools/include/plugin.hpp is not really very good design, but it also avoid the "kdb" prefixes.

Instead of the modules as parameter it is nicer to have a PluginLoader, which is a very usual thing in Java.

markus2330 replied:
Furthermore, it would be nice if one gets a "Plugin" in both cases, as the NativePlugin and the current "Plugin" obviously should be subclass of a common class.

Piankero replied:
> It is astonishing that such important decisions are not mentioned at all. Please document such things, it would have saved a lot of time for both of us.

Why is it an important decision how I name a method? I will simply revert it to the  non-prefix version.

> Instead of the modules as parameter it is nicer to have a PluginLoader, which is a very usual thing in Java.

You mean you want some sort of "Factory"? The current NativePlugin is just a proxy to abstract the pointer logic away and implement the Plugin Interface correctly.

> Furthermore, it would be nice if one gets a "Plugin" in both cases, as the NativePlugin and the current "Plugin" obviously should be subclass of a common class.

Currently the `Plugin` class is an interface and the NativePlugin along with the other plugins implement it. I do not know what you mean here.

markus2330 replied:
>  Why is it an important decision how I name a method? I will simply revert it to the non-prefix version.

It is not only about the method names: it was completely unclear me what the whole NativeElektraPlugin is about. The renaming of the public methods is also problematic: it breaks all already written plugins.

> You mean you want some sort of "Factory"?

Yes, something like https://docs.oracle.com/javase/7/docs/api/java/lang/ClassLoader.html

> The current NativePlugin is just a proxy to abstract the pointer logic away and implement the Plugin Interface correctly.

Which does not say anything about if this design is good.

>  Currently the Plugin class is an interface and the NativePlugin along with the other plugins implement it. I do not know what you mean here.

I was talking about the ClassLoader: it could be a secret that such a thing like a NativePlugin exists at all.

Piankero replied:
I am still not sure how exactly how I should name the NativeElektraPlugin as it has to have a name. I can make a PluginLoader but still I need to have name for the class that loads the NativePlugins from elektra and not those implemented in java

markus2330 replied:
NativePlugin or Plugin in another package should be ok.

markus2330 commented on src/bindings/jna/libelektra4j/src/main/java/org/libelektra/exception/ConflictingStateException.java:
```diff
@@ -0,0 +1,16 @@
+
+public class ConflictingStateException extends KDBException {
+
+	private static String errorCode = "C02000";
```
final?

markus2330 commented on src/bindings/jna/libelektra4j/src/main/java/org/libelektra/exception/KDBException.java:
```diff
@@ -0,0 +1,90 @@
+import java.util.Collection;
+
+/**
+ * Custom KDB exception class being used for I/O errors
```
for I/O errors?

Piankero replied:
This documentation was copied from the old Exception class which I patched. I was also wondering why `I/O` errors are meant but thought it will have some special meaning behind it. I can remove it

markus2330 replied:
Please do not copy text you do not understand. I do not think that any part of that text is correct for what you implemented here. (It is not "Custom", and not used for I/O errors.)

Piankero replied:
> Please do not copy text you do not understand. I do not think that any part of that text is correct for what you implemented here. (It is not "Custom", and not used for I/O errors.)

Neither was it before then obviously but you usually complain if I delete any comments so I kept them in case the author before me know more than I did. I will remove them

markus2330 replied:
Because both is wrong:

1. having wrong comments
2. having no comments

Please properly document your methods according design-by-contract.

markus2330 commented on src/bindings/jna/libelektra4j/src/main/java/org/libelektra/exception/KDBException.java:
```diff
@@ -0,0 +1,90 @@
+ */
+public abstract class KDBException extends java.io.IOException {
+
+	private static final long serialVersionUID = 1L;
```
?

Piankero replied:
Same reason as above

markus2330 replied:
Where above?

Piankero replied:
The previous author added this for serialization versioning (which was also not very intuitive for me). I will just delete it

markus2330 replied:
Yes, I do not know which use case the previous author had in mind for serializing the errors.

markus2330 commented on src/bindings/jna/libelektra4j/src/main/java/org/libelektra/exception/KDBException.java:
```diff
@@ -0,0 +1,90 @@
+		return errorKey;
+	}
+
+	public String getErrorCode() {
```
There is a huge mess with code/number. Please use always the same.

Piankero replied:
What do you mean with "mess"? Do I sometimes write "code" and sometimes "number"?

markus2330 replied:
Yes, see the line below.

markus2330 commented on src/bindings/jna/libelektra4j/src/main/java/org/libelektra/exception/LogicalException.java:
```diff
@@ -0,0 +1,16 @@
+		super(k);
+	}
+
+	public static String errorCode() {
```
Why no @Override here?

Piankero replied:
Because it is a static method?

markus2330 replied:
Exactly, wouldn't that possibly return wrong errorCodes in catch blocks?

Piankero replied:
Why should it? The error code can be returned via `e.getErrorNumber()` which is in the KDBException and actually calls `errorKey.getMeta("error/number").getString()`. The static method is never used in a catch block

markus2330 replied:
It would be far more efficient if you improve the docu/code and avoid starting discussions. (Do you really want me to explain why someone might invoke a public method?)

Anyway: If a user should not use the method, why is it there, then? Docu does not give a glue. Furthermore, we again have a mess in code/number.

Piankero replied:
It would not even come into my mind that something like this should be documented because it is a standard java coding style.

```java
try {
//code that causes exception
} catch (KDBException e) { \\or any subexception
	e.getErrorNumber();
}
```
Why should somebody call a static method here?

The static method is just useful so one does not have to provide a no-arg constructor (which would even be misleading and not even be possible for the abstract classes) and use `new XYException().getErrorNumber()` or lookup the error number in the design decisions.

markus2330 commented on src/bindings/jna/libelektra4j/src/main/java/org/libelektra/plugin/NativeElektraPlugin.java:
```diff
@@ -0,0 +1,169 @@
+/**
+ * This class can be used to load native C/C++ written Elektra Plugins to be used by Java directly
+ */
+public class NativeElektraPlugin implements Plugin {
```
See above, another name would fit better.

markus2330 commented on src/bindings/jna/libelektra4j/src/main/java/org/libelektra/plugin/NativeElektraPlugin.java:
```diff
@@ -0,0 +1,169 @@
+import java.util.List;
+
+/**
+ * This class can be used to load native C/C++ written Elektra Plugins to be used by Java directly
```
```suggestion
 * This class can be used to load native Elektra Plugins in Java directly
```

markus2330 commented on src/bindings/jna/libelektra4j/src/test/java/org/libelektra/ErrorCodeTest.java:
```diff
@@ -0,0 +1,77 @@
+		}
+		throw new RuntimeException("Exception did not trigger");
+	}
+}
```
Not many tests here. No newline at the end.

Piankero replied:
It tests the basic functionality of the errors with success. Which other tests do you have in mind?

markus2330 replied:
There are so many things not tested (related to errors). I somewhat understand that you do not want to make so many tests in C and shell but that you also write so little tests in Java is surprising for me. As you removed the asserts in two tests, you literally wrote a single test. If you really do not know what to test: You can easily find them via code coverage.  E.g. warnings, warnings/errors from native plugins, ... are obvious parts.

Piankero commented on src/bindings/jna/README.md:
```diff
@@ -113,6 +113,64 @@ Most likely, you have to include this dependency as well:
 
+### Using Elektra plugins directly
+
+The `NativeElektraPlugin` can be used to load a native Elektra plugin which is written in C/C++.
```
I don't think "loader" is correct as the class itself then acts as the plugin.

markus2330 replied:
It is also weird that both things are married in one class. See in other comment about design.

Piankero commented on src/bindings/jna/libelektra4j/src/main/java/org/libelektra/KeySet.java:
```diff
@@ -43,7 +43,7 @@ protected KeySet(final long p) {
 	 * @param p Pointer to another KeySet
 	 */
-	protected KeySet(final Pointer p) {
+	public KeySet(final Pointer p) {
```
Because I moved the NativeElektraPlugin into the plugin package and it makes usage of this method. It does not harm the development experience though if this is public.

Piankero replied:
I did not check this, I can throw a RunTime Exception if the pointer is null but apart from that I do not see any other way around making the constructor public so that the NativePlugin can load

markus2330 replied:
Not ideal but if your design needs it you can add the check. Please document invariants.

markus2330 commented on src/bindings/jna/README.md:
````diff
@@ -113,6 +113,64 @@ Most likely, you have to include this dependency as well:
+
+```java
+Key parentKey = Key.create("user/tests/javabinding");
+String errorPluginName = "error";
````
But only once?

Piankero commented:
@markus2330 could you please take a look now. I will update the Readme and test cases once you agree on the core structure of the java binding (PluginLoader, method naming, etc.)

markus2330 commented:
I do not know what to say, most of what was discussed is still not fixed.

Piankero commented:
> I do not know what to say, most of what was discussed is still not fixed.

1. I added a `PluginLoader`
2. I renamed everything from `errorCode` to `errorNumber`
3. I added javadoc to nearly all methods

What exactly is not fixed listing? I am not yet 100% sure of your plugin loader implementation wish so I asked you to take a look at it now

markus2330 commented on src/bindings/jna/libelektra4j/src/main/java/org/libelektra/PluginLoader.java:
```diff
@@ -0,0 +1,40 @@
+
+/**
+ * This class can be used to load Plugins from Elektra.
+ * It also loads self implemented Java plugins.
```
Here I would have the expectation that also my newly implemented Java plugin can be loaded (which obviously is not the case as the loading is hard-coded).

Piankero replied:
I could use reflection but this comes with several drawbacks.
From java 9 on Reflection is not very favorable anymore
Furthermore one would need to force a default constructor so every plugin can be instantiated via reflection.

I will add it to the README that it also has to be added to the PluginLoader implementation.

markus2330 replied:
Yes, I also think the readme how to add the plugin would be enough.

What is not so ideal in the current solution that the Java plugins shadow other plugins. If there is also a native "echo" plugin, it would not be possible to load it.

Maybe we drop the idea to "load" Java plugins altogether.

Piankero replied:
> What is not so ideal in the current solution that the Java plugins shadow other plugins. If there is also a native "echo" plugin, it would not be possible to load it.

I agree with you. I separate the methods into a "loadNativePlugin" and "loadJavaPlugin" to make it clear.

markus2330 replied:
good idea :+1:

markus2330 commented on src/bindings/jna/libelektra4j/src/main/java/org/libelektra/PluginLoader.java:
```diff
@@ -0,0 +1,40 @@
+	}
+
+	/**
+	 * This plugin loads a Java/Native Elektra Plugin.
```
```suggestion
	 * This class loads a Java or native Elektra plugins.
```

markus2330 commented on src/bindings/jna/libelektra4j/src/main/java/org/libelektra/PluginLoader.java:
```diff
@@ -0,0 +1,40 @@
+	 * @throws InstallationException if the plugin does not exist
+	 */
+	public Plugin loadPlugin(String name) throws InstallationException {
+		if (name.equals(new Echo().getName())) {
```
I am not sure we need direct instantiation of Java plugins. But if we do it should be done in a proper way. The current implementation instantiated every plugin, even if for none of them the name matches.

markus2330 commented on doc/news/_preparation_next_release.md:
```diff
@@ -98,7 +98,9 @@ you up to date with the multi-language support provided by Elektra.
 
-- Upgraded maven dependencies for java binding _(Michael Zronek)_
+- Upgraded maven dependencies for Java binding _(Michael Zronek)_
+- Completely overhauled the Java binding to be able to use Elektra plugins directly. For an example see the [test case](../../src/bindings/jna/libelektra4j/src/test/java/org/libelektra/ErrorCodeTest.java). _(Michael Zronek)_
```
Here it is still called ErrorCode

markus2330 commented on src/bindings/jna/libelektra4j/src/main/java/org/libelektra/PluginLoader.java:
```diff
@@ -0,0 +1,40 @@
+
+	private Key errorKey;
+
+	public PluginLoader(Key errorKey) {
```
Javadoc still missing. Why is this constructor needed? You can also instantiate the errorKey internally?

Piankero replied:
Developers should also have the possibility to define the errorKey themselve in my opinion to access its properties.

Should I add a no-arg constructor which with the default error key name `system/jna/default/errorkey`?

markus2330 replied:
> Should I add a no-arg constructor which with the default error key name system/jna/default/errorkey?

Yes, but I would leave the name empty.

markus2330 commented on src/bindings/jna/libelektra4j/src/main/java/org/libelektra/PluginLoader.java:
```diff
@@ -0,0 +1,40 @@
+		if (name.equals(new Return().getName())) {
+			return new Return();
+		}
+		return new NativePlugin(name, errorKey);
```
One more important thing: The PluginLoader should be responsible for the modules. The NativePlugin should not reopen the modules every time (such a constructor should not be offered as it would be unnecessarily slow).

PhilippGackstatter commented on src/bindings/jna/libelektra4j/src/main/java/org/libelektra/exception/mapper/ExceptionMapperService.java:
```diff
@@ -0,0 +1,49 @@
+			temporaryError.setMeta("error/number", InternalException.errorNumber());
+			temporaryError.setMeta("error/reason", "Sorry, Elektra returned an error/warning but no " +
+					"information along it, please report this incident at https://issues.libelektra.org/");
+			return new InternalException(temporaryError);
```
Just a question here: If you return this exception and later someone calls `exception.getModule()`, would that fail? Because the `error/module` key is not set here and `getModule` doesn't do any checks and doesn't throw exceptions either: 
```java
public String getModule() [ getModule()]  {
		return errorKey.getMeta("error/module").getString();
	}
```

I'm implementing the kdb errors for the Rust binding atm, so I'm looking at this PR for some inspiration and noticed that. So to fully handle the case where an error key returned from a kdb function is empty, shouldn't all the `error/*` metakeys be set?

markus2330 replied:
Yes, I also found the code weird, see my comments two lines above.

From my perspective, the `kdb*` calls must properly set the errors. It would be wrong to fix problems from the core in every binding separately.

> So to fully handle the case where an error key returned from a kdb function is empty, shouldn't all the error/* metakeys be set?

This case does not need to be handled, as it would be a bug in the core and needs to be fixed there. And it is actually quite simple to fix all such errors: for every `return -1` in the `kdb` functions, we simply need to check if an error was correctly set, and set some error (similar to the code above) if it was not set.

Piankero commented:
@markus2330 the build fails: 
```genhtml: ERROR: cannot write HTML data (No space left on device)```

Piankero commented:
jenkins build libelektra please

markus2330 commented:
The build server fails, so it cannot be "ready to merge".

Piankero commented:
jenkins build libelektra please

Piankero commented:
The build server fails quite often and just hang on the last step which was docker related (https://build.libelektra.org/jenkins/blue/organizations/jenkins/libelektra/detail/PR-2854/44/pipeline/). I think it is still ready to merge because all tests ran through

markus2330 commented on src/bindings/jna/libelektra4j/src/main/java/org/libelektra/Key.java:
```diff
@@ -489,10 +469,10 @@ public int copyMeta(final Key source, final String metaName) {
-	 * @return 1 if meta was successfully copied, 0 if source doesn't contain any meta and nothing had to be done, -1 in case of an error or
-	 *         if the source parameter was null
+	 * @param source Key object that is used as source
+	 * @return 1 if meta was successfully copied, 0 if source doesn't contain any meta and nothing had to be done, -1
```
should be @retval

Piankero replied:
@*RetVal* is not a standard javadoc annotation (https://www.oracle.com/technetwork/java/javase/documentation/index-137868.html). Do you still want it nonetheless?

markus2330 replied:
This is parsed by doxygen, so I do not see why javadoc is relevant here. Or do you want to build this with javadoc instead?

markus2330 commented on src/bindings/jna/libelektra4j/src/main/java/org/libelektra/Key.java:
```diff
@@ -472,12 +453,11 @@ public Key currentMeta() {
-	 *            Key name of the meta to be copied
-	 * @return 1 if meta was successfully copied, 0 if source doesn't contain the required meta and nothing had to be done, -1 in case of an
-	 *         error or if the source parameter was null
+	 * @param source   Key object that is used as source
```
Did you try to build the docu? Does it look nice? https://doc.libelektra.org/api/current/html/classorg_1_1libelektra_1_1KeySet.html

Piankero replied:
To me it looks nice:

markus2330 commented on src/bindings/jna/libelektra4j/src/main/java/org/libelektra/Key.java:
```diff
@@ -472,12 +453,11 @@ public Key currentMeta() {
-	 *         error or if the source parameter was null
+	 * @param source   Key object that is used as source
+	 * @param metaName Key name of the meta to be copied
+	 * @return 1 if meta was successfully copied, 0 if source doesn't contain the required meta and nothing had to be
```
should be @retval here

markus2330 commented on src/bindings/jna/libelektra4j/src/main/java/org/libelektra/KeySet.java:
```diff
@@ -40,10 +40,14 @@ protected KeySet(final long p) {
-	protected KeySet(final Pointer p) {
+	public KeySet(final Pointer p) {
+		if (p == null) {
+			throw new IllegalArgumentException("Passed pointer may not be null");
```
Should be NullPointerException?

```suggestion
			throw new IllegalArgumentException("Passed pointer should not be null");
```

markus2330 commented on src/bindings/jna/libelektra4j/src/main/java/org/libelektra/plugin/NativePlugin.java:
```diff
@@ -0,0 +1,182 @@
+	 * @throws InstallationException if the plugin does not exist
+	 */
+	public NativePlugin(String pluginName, Key errorKey) throws InstallationException {
+		KeySet modules = KeySet.create();
```
modules are still recreated every time, this should be in the module loader.

markus2330 commented:
> The build server fails quite often

Yes, please improve #2855 

> think it is still ready to merge because all tests ran through

There are also later checks (webui, Debian packages, ...).

I'll merge it for now. Please further improve the JNA binding in a new PR.
